lld/ELF/InputFiles.cpp

```diff
@@ -1418,6 +1439,38 @@ std::vector<uint32_t> SharedFile::parseVerneed(const ELFFile<ELFT> &obj,
   return verneeds;
 }
 
+// To determine if a shared file can support any of the GNU Attributes,
+// the .note.gnu.properties section need to be read. The appropriate
+// location in memory is located then the GnuPropertyNote can be parsed.
+// This is the same process as is used for readGnuProperty, however we
+// do not pass the data variable as, without an InputSection, its value
+// is unknown in a SharedFile. This is ok as the information that would
+// be collected from this is irrelevant for a dynamic object.
+template <typename ELFT>
+void SharedFile::parseGnuAndFeatures(const uint8_t *base,
+                                     const typename ELFT::PhdrRange headers,
+                                     const typename ELFT::Shdr *sHeader) {
+  if (numElfPhdrs == 0 || sHeader == nullptr)
```

Comment:
The program header and the section header are describing the same part of the ELF file.

For a shared object there is always a program header but there may not be a section header table (if the shared-library has been stripped).

It looks like the code only uses sHeader to get the section alignment. You can get this from the p_align field of the program header, then you don't need the sHeader at all.